service/history/api/updateworkflow/api.go

```diff
@@ -132,12 +131,21 @@ func (u *Updater) ApplyRequest(
 	updateReg update.Registry,
 	ms workflow.MutableState,
 ) (*api.UpdateWorkflowAction, error) {
+	updateID := u.req.GetRequest().GetRequest().GetMeta().GetUpdateId()
+
 	if !ms.IsWorkflowExecutionRunning() {
+		// If the WF is not running anymore, use an existing Update, if it exists for the requested ID.
+		// This ensures that repeated Update requests with the same ID see the same result.
+		if u.upd = updateReg.Find(ctx, updateID); u.upd != nil {
+			return &api.UpdateWorkflowAction{Noop: true}, nil
+		}
 		return nil, consts.ErrWorkflowCompleted
 	}
+
 	u.wfKey = ms.GetWorkflowKey()
 
-	if u.req.GetRequest().GetFirstExecutionRunId() != "" && ms.GetExecutionInfo().GetFirstExecutionRunId() != u.req.GetRequest().GetFirstExecutionRunId() {
+	if u.req.GetRequest().GetFirstExecutionRunId() != "" &&
+		ms.GetExecutionInfo().GetFirstExecutionRunId() != u.req.GetRequest().GetFirstExecutionRunId() {
```

Comment:
I believe this check needs to be done before `!ms.IsWorkflowExecutionRunning()` check now.

Reply:
👍 agreed!